### PR TITLE
Fix: Bus didn't send events to all receivers

### LINKED ISSUE #58: Horust hangs forever when received SIGTERM before all services go up and running.
#### Version Information:

* Version: v0.1.3

First noticed the bug on my modified branch (see PR-draft #57), however, the same bug occurs on fresh `master` and when my friend installed `horust` via `cargo install`.

* OS: Linux
* CPU: x64

#### Description:
Scenario:

* I start horust services
* Not all services are running, at least one is still "Starting",
* I have to quickly kill them all (because for example I made a mistake in configuration and I don't want to wait till service goes up)

Then Horust will terminate all already running services, but waits forever for this late service which is never changing its state from `Starting` to `Started`.

It means that in the main loop Horust checks whether all services are `Finished` or `Failed`, but this one service will just wait forever in `Starting` state. I guess for some reason it never receives `SpawnFailed` event?

The quickest way to test it is to add `start-delay = "2s"` in `service.toml`.

I made some `eprintln!` and tried to debug it:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
[2021-04-29T07:51:21Z INFO  horust] Loading services from directories:
    * ./core
    * ./extra
[2021-04-29T07:51:21Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:21Z INFO  horust::horust::supervisor] Applying events... [
    Run("1.toml"), Run("2.toml"), Run("3.toml"), Run("4.toml"), Run("5.toml"), Run("6.toml"), Run("7.toml"), Run("8.toml"), Run("9.toml"), Run("10.toml")
]
[2021-04-29T07:51:22Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:22Z INFO  horust::horust::supervisor] Applying events... []
^C[2021-04-29T07:51:22Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:22Z WARN  horust::horust::supervisor] 1. SIGTERM received
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: SpawnFailed("6.toml")
[2021-04-29T07:51:22Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:22Z WARN  horust::horust::supervisor] 1. SIGTERM received
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: SpawnFailed("7.toml")
[2021-04-29T07:51:23Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:23Z WARN  horust::horust::supervisor] 1. SIGTERM received
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: SpawnFailed("8.toml")
[2021-04-29T07:51:23Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:23Z WARN  horust::horust::supervisor] 1. SIGTERM received
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: SpawnFailed("4.toml")
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("2.toml", Pid(66623))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("5.toml", Pid(66624))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("1.toml", Pid(66625))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("10.toml", Pid(66626))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("3.toml", Pid(66627))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("9.toml", Pid(66628))
[2021-04-29T07:51:23Z INFO  horust::horust::supervisor] Applying events... [
    PidChanged("2.toml", Pid(66623)),
    PidChanged("5.toml", Pid(66624)),
    PidChanged("1.toml", Pid(66625)),
    PidChanged("10.toml", Pid(66626)),
    PidChanged("3.toml", Pid(66627)),
    PidChanged("9.toml", Pid(66628))
]
[2021-04-29T07:51:23Z WARN  horust::horust::supervisor] 1. SIGTERM received
PID CHANGED: "2.toml" - 66623
PID CHANGED: "5.toml" - 66624
PID CHANGED: "1.toml" - 66625
PID CHANGED: "10.toml" - 66626
PID CHANGED: "3.toml" - 66627
PID CHANGED: "9.toml" - 66628
[2021-04-29T07:51:24Z INFO  horust::horust::supervisor] Applying events... [
    ShuttingDownInitiated(Gracefuly),
    StatusChanged("2.toml", Started),
    StatusChanged("5.toml", Started),
    StatusChanged("1.toml", Started),
    StatusChanged("10.toml", Started),
    StatusChanged("3.toml", Started),
    StatusChanged("9.toml", Started)
]
[2021-04-29T07:51:24Z WARN  horust::horust::supervisor] 1. SIGTERM received
[2021-04-29T07:51:24Z WARN  horust::horust::supervisor] Gracefully stopping...
[2021-04-29T07:51:24Z INFO  horust::horust::supervisor] Applying events... [
    ShuttingDownInitiated(Gracefuly),
    StatusUpdate("1.toml", InKilling),
    Kill("1.toml"),
    StatusUpdate("2.toml", InKilling),
    Kill("2.toml"),
    StatusUpdate("3.toml", InKilling),
    Kill("3.toml"),
    StatusUpdate("5.toml", InKilling),
    Kill("5.toml"),
    StatusUpdate("11.toml", Finished),
    StatusUpdate("9.toml", InKilling),
    Kill("9.toml"),
    StatusUpdate("10.toml", InKilling),
    Kill("10.toml")
]
[2021-04-29T07:51:24Z WARN  horust::horust::supervisor] Gracefully stopping...
[2021-04-29T07:51:24Z INFO  horust::horust::supervisor] Applying events... [
    StatusChanged("1.toml", InKilling),
    StatusChanged("2.toml", InKilling),
    StatusChanged("3.toml", InKilling),
    StatusChanged("5.toml", InKilling),
    StatusChanged("11.toml", Finished),
    StatusChanged("9.toml", InKilling),
    StatusChanged("10.toml", InKilling)
]
[2021-04-29T07:51:25Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:25Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:25Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:25Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:26Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:26Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:26Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:27Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:27Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:27Z INFO  horust::horust::supervisor] Applying events... [
    ForceKill("1.toml"),
    ForceKill("2.toml"),
    ForceKill("3.toml"),
    ForceKill("5.toml"),
    ForceKill("9.toml"),
    ForceKill("10.toml")
]
[2021-04-29T07:51:28Z INFO  horust::horust::supervisor] Applying events... [
    StatusChanged("1.toml", Failed),
    StatusChanged("2.toml", Failed),
    StatusChanged("3.toml", Failed),
    StatusChanged("5.toml", Failed),
    StatusChanged("9.toml", Failed),
    StatusChanged("10.toml", Failed),
    StatusUpdate("1.toml", FinishedFailed),
    StatusUpdate("2.toml", FinishedFailed),
    StatusUpdate("3.toml", FinishedFailed),
    StatusUpdate("5.toml", FinishedFailed),
    StatusUpdate("9.toml", FinishedFailed),
    StatusUpdate("10.toml", FinishedFailed)
]
[2021-04-29T07:51:28Z INFO  horust::horust::supervisor] Applying events... [
    StatusChanged("1.toml", FinishedFailed),
    StatusChanged("2.toml", FinishedFailed),
    StatusChanged("3.toml", FinishedFailed),
    StatusChanged("5.toml", FinishedFailed),
    StatusChanged("9.toml", FinishedFailed),
    StatusChanged("10.toml", FinishedFailed)
]
[2021-04-29T07:51:28Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:28Z INFO  horust::horust::supervisor] Applying events... []
At this point there are no new events forever (loop with empty Applying events... []
```
My modifications:
* Added Graceful/Forceful stop - but as I wrote, it isn't the cause as the bug occurs also on fresh `main` branch.
* Added `SPAWN_FOR_EXEC_HANDLER_LOOP` eprintln! in `process_spawner` just after `loop` ends (before `bus.send_event(ev)`. Thanks to that, we can see that `SpawnFailed` event is created!
* I added two extra `exprintln!`: in `supervisor/mod.rs::handle_event()` to see if we ever process `SpawnFailed` or `PidChanged`. One is processed, another isnt.
* So, I changed log level for `Applying events` to `info` -> We can see that `SpawnFailed` event is never received!
* But `PidChanged` is received. Why?

Extra conclusion:
Whats interesting, one service ("11.toml") had `start-delay = 10s` instead of `2s`, and it had `start-after=["8.toml"]`.
So horust correctly were waiting till `8.toml` goes up, but its hanged on `Starting` (Spawn Failed has not been processed), therefore `11.toml` is in `Initial` state, so `service_handler.rs` changed its status to `Finished` correctly.

The only way to kill horust then is `killall horust -9`

### PR DESCRIPTION
**Pull request by frondeus**
### Motivation and Context

fixes #58 

### Description
In few places there was a `.clone()` on the `BusConnector<T>`.
Unfortunately the `.clone()` created a round-robin kind of message sending instead of the broadcast.
It means that not every thread would receive all messages.

I noticed that any thread can request a `join_bus()`, even after the application startup.
It required creating an `Arc<Mutex<>>` on a list of all senders because this list can be modified in a runtime.
It means there is some performance loss, but I believe correctness is more important.

### How Has This Been Tested?

I tested it manually, see #58. But also I wrote a unit test to prove that `Bus` is now working as intended - every time in broadcast mode.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: src/horust/bus.rs [hunk 1]
```diff
@@ -6,8 +6,41 @@
 //!
 
 use crossbeam::channel::{unbounded, Receiver, Sender};
-use std::fmt::Debug;
 use std::fmt::Formatter;
+use std::{
+    fmt::Debug,
+    sync::{Arc, Mutex},
+};
+
```
Comment (FedericoPonzi) [inline src/horust/bus.rs:14]: Can you add few words of documentation?

File: src/horust/bus.rs [hunk 2]
```diff
@@ -6,8 +6,41 @@
...
+    /// Add another connection to the bus
+    pub fn join_bus(&self) -> BusConnector<T> {
+        let mut senders = self.senders.lock().unwrap();
+
+        let (sender, receiver) = unbounded();
+
+        let id = senders.len() as u64;
+        senders.push((id, sender));
+
+        drop(senders);
```
Comment (FedericoPonzi) [inline src/horust/bus.rs:39]: I'm curious, why do you need the explicit drop? to release the lock before the self.clone?

Comment (frondeus) [reply src/horust/bus.rs]: >  to release the lock before the self.clone?

Well, yes.  That was my intuition - however, I checked on rust playground, and it's not necessary.

Comment (FedericoPonzi) [reply src/horust/bus.rs]: Cool, I knew that locks are released as soon as they go out of scope, but I wasn't 100% sure with the self.clone and an Arc, I thought that maybe weird things could happen without explicit dropping before cloning...

**Pull request merged**